### PR TITLE
Implementation of smoothing stabilization

### PR DESCRIPTION
**Pull request by rrsadykov**

### CODE REVIEW

**Comment by rrsadykov** [comment]
@guimarqu What is the best way to do a linear combination of two `DualSolution`? For example, if we have solutions (constr_1 = 2, constr_2 = 2) and (constr_1 = 1, constr_3 = 1), then linear combination with coefficient 0.5 should be (constr_1 = 1.5, constr_2 = 1, constr_3 = 0.5). In addition, the coefficient should depend on the constraint type (coefficient should be different for convexity constraints).

**Comment by guimarqu** [comment]
> @guimarqu What is the best way to do a linear combination of two `DualSolution`? For example, if we have solutions (constr_1 = 2, constr_2 = 2) and (constr_1 = 1, constr_3 = 1), then linear combination with coefficient 0.5 should be (constr_1 = 1.5, constr_2 = 1, constr_3 = 0.5). In addition, the coefficient should depend on the constraint type (coefficient should be different for convexity constraints).  

If the coefficient depends on the duty, it’s better to use a loop.

**Comment by rrsadykov** [comment]
If the coefficient is independent, is it possible to perform the linear combination in O(n)?
Because if one uses a loop, the complexity is O(n\log n), no?
I think that the best is to do the linear combination with the same coefficient in O(n), and then make another loop in O(n) to correct the coefficients of convexity constraints.

**Comment by guimarqu** [comment]
Loop is O(n), lookup is O(log n).
If you do something like that, you are in O(n) :

```julia
constrids = [constrid for (constrid, _) in getconstr(form)] # sorted ? If not, we should consider keeping the constraint dict sorted
constrvals = zeros(Int, length(constrids)

pos = 1
for (constrid, constrval) in dualsolution # sorted and O(n)
    while constrids[pos] < constrid
         pos += 1
    end
    if ! (constrids[pos] > constrid)
        constrvals[pos] += coefficient * constrval
    end
end
```

**Comment by guimarqu** [comment]
It's not a good idea to keep the constraints dict sorted. If we assume that no constraints are added during column generation, we can sort the constrids once at the beginning of colgen.

**Comment by rrsadykov** [comment]
Ok, I will first do a simple version in O(n\log n). We will see latter if we need to improve that

**Comment by rrsadykov** [comment]
@guimarqu In `Solution.sol`, `ConstrId` are sorted or not?

**Comment by guimarqu** [comment]
> @guimarqu In `Solution.sol`, `ConstrId` are sorted or not? 

Yes, they are.

**Comment by rrsadykov** [comment]
@guimarqu Guillaume, can you review and merge? All tests without stabilization work.
I prefer to do an "intermediate" merge now as there are many changes in column generation algorithm and it will simplify merging with other branches in the future. 

For the moment, stabilization is not quite ready yet, but the column generation algorithm is fully prepared for it. I think I need a couple days more to finish the implementation and add tests. 

Can you leave the branch after merge so I can continue

**Comment by guimarqu** [comment]
Do you want to finish stabilization before release of 0.3 (we can wait until mid-June max) ?

File: src/Algorithm/colgen.jl [hunk 1]
```diff
@@ -128,54 +121,94 @@ function should_do_ph_1(master::Formulation, data::ColGenRuntimeData)
...
+    pseudo_dual_bound_contrib::Float64
+    recorded_sol_ids::Vector{VarId}
+    sol_ids_to_activate::Vector{VarId}
+    isfeasible::Bool
+end
+
+function SubprobInfo(reform::Reformulation, spformid::FormId)
+    master = getmaster(reform)
+    lb_constr_id = reform.dw_pricing_sp_lb[spformid]
+    ub_constr_id = reform.dw_pricing_sp_ub[spformid]
```
[the remark this answers is not in the document]
Comment (guimarqu) [reply src/Algorithm/colgen.jl:180]: new issue #366

File: src/Algorithm/colgen.jl [hunk 2]
```diff
@@ -9,7 +9,7 @@
         max_nb_iterations::Int = 1000
         optimality_tol::Float64 = 1e-5
         log_print_frequency::Int = 1
-        store_all_ip_primal_sols::Bool = false
+        store_all_ip_primal_sols::Bool = falserecorded_sol_ids
```
[the remark this answers is not in the document]
Comment (rrsadykov) [reply src/Algorithm/colgen.jl:12]: yes

File: src/Algorithm/colgen.jl [hunk 3]
```diff
@@ -128,54 +121,94 @@ function should_do_ph_1(master::Formulation, data::ColGenRuntimeData)
...
         else
             setcurcost!(master, varid, getperencost(master, var))
         end
     end
-    set_lp_dual_bound!(data.optstate, DualBound(master))
-    set_ip_dual_bound!(data.optstate, DualBound(master))
-    data.phase = 2
+    set_lp_dual_bound!(optstate, DualBound(master))
+    set_ip_dual_bound!(optstate, DualBound(master))
 end
```
[the remark this answers is not in the document]
Comment (rrsadykov) [reply src/Algorithm/colgen.jl:145]: ok

File: src/Algorithm/colgen.jl [hunk 4]
```diff
@@ -184,55 +217,88 @@ function insert_cols_in_master!(
...
+            end 
+        end
+    else
+        red_cost = getvalue(spsol)    
+    end
+    #println("Red cost before conv. constr. duals is ", red_cost)
+    if getobjsense(master) == MinSense
+        red_cost -= spinfo.lb * spinfo.lb_dual + spinfo.ub * spinfo.ub_dual
+    else
+        red_cost += spinfo.lb * spinfo.lb_dual + spinfo.ub * spinfo.ub_dual
```
[the remark this answers is not in the document]
Comment (rrsadykov) [reply src/Algorithm/colgen.jl:274]: ok

File: src/Algorithm/colgen.jl [hunk 5]
```diff
@@ -128,54 +121,94 @@ function should_do_ph_1(master::Formulation, data::ColGenRuntimeData)
...
+    spinfo.isfeasible = true
+    empty!(spinfo.recorded_sol_ids)
+    empty!(spinfo.sol_ids_to_activate)
+end
+
+set_bestcol_id!(spinfo::SubprobInfo, varid::VarId) = spinfo.bestcol_id = varid
+
+add_recorded_sol_id!(spinfo::SubprobInfo, varid::VarId) = push!(spinfo.recorded_sol_ids, varid)
+
+add_sol_id_to_activate!(spinfo::SubprobInfo, varid::VarId) = push!(spinfo.sol_ids_to_activate, varid)
```
[the remark this answers is not in the document]
Comment (rrsadykov) [reply src/Algorithm/colgen.jl:203]: ok, removed

File: src/Algorithm/colgen.jl [hunk 6]
```diff
@@ -349,124 +415,88 @@ function updatereducedcosts!(reform::Reformulation, redcostsvec::ReducedCostsVec
...
-            gen_status, new_sp_sol_ids, sp_sol_ids_to_activate, sp_dual_contrib = solve_sp_to_gencol!(
-                algo, masterform, spdata, dual_sol, sp_lbs[spuid], sp_ubs[spuid]
-            )
-            gen_statuses[spuid] = gen_status
-            if gen_status # else Sp is infeasible: contrib = Inf
-                recorded_sp_solution_ids[spuid] = new_sp_sol_ids
-                sp_solution_to_activate[spuid] = sp_sol_ids_to_activate
-            end
-            sp_dual_bound_contribs[spuid] = sp_dual_contrib #float(contrib)
+            solve_sp_to_gencol!(algo, masterform, spdata, lp_dual_sol, spinfos[spuid])
```
[the remark this answers is not in the document]
Comment (rrsadykov) [reply src/Algorithm/colgen.jl:435]: ok, changed

File: src/Algorithm/colgen.jl [hunk 7]
```diff
@@ -349,124 +415,88 @@ function updatereducedcosts!(reform::Reformulation, redcostsvec::ReducedCostsVec
...
-        update_lp_dual_bound!(cg_optstate, lagran_bnd)
-        if nb_new_col < 0
-            # subproblem infeasibility leads to master infeasibility
-            return -1
-        end
-        break # TODO : rm
-    end
-    return nb_new_columns
-end
+# function generatecolumns!(
```
[the remark this answers is not in the document]
Comment (rrsadykov) [reply src/Algorithm/colgen.jl:467]: deleted

File: src/Algorithm/colgen.jl [hunk 8]
```diff
@@ -496,29 +526,92 @@ function cleanup_columns(algo::ColumnGeneration, nb_iterations::Int64, data::Ref
...
+        spinfo.ub_dual = dualsol[spinfo.ub_constr_id]    
+        dualsol[spinfo.lb_constr_id] = zero(0.0)
+        dualsol[spinfo.ub_constr_id] = zero(0.0)
+        if getobjsense(form) == MinSense
+            newbound -= spinfo.lb_dual * spinfo.lb 
+            newbound -= spinfo.ub_dual * spinfo.ub 
+        else
+            newbound += spinfo.lb_dual * spinfo.lb 
+            newbound += spinfo.ub_dual * spinfo.ub 
+        end
```
[the remark this answers is not in the document]
Comment (rrsadykov) [reply src/Algorithm/colgen.jl:587]: ok

File: src/Algorithm/colgenstabilization.jl [hunk 9]
```diff
@@ -0,0 +1,242 @@
...
+        storage.pseudo_dual_bound = pseudo_lagr_bound
+    end
+end
+
+function update_stab_after_colgen_iteration!(storage::ColGenStabStorage)
+    if storage.newstabcenter !== nothing
+        storage.stabcenter = storage.newstabcenter
+    end
+    storage.curalpha = 0.0 
+    storage.newstabcenter = nothing
```
[the remark this answers is not in the document]
Comment (rrsadykov) [reply src/Algorithm/colgenstabilization.jl:241]: corrected

**Comment by rrsadykov** [comment]
I think I can finish this week

File: src/Algorithm/colgenstabilization.jl (hunk 9 above)
[the remark this answers is not in the document]
Comment (guimarqu) [reply src/Algorithm/colgenstabilization.jl:241]: I don't see it.

[the remark this answers is not in the document]
Comment (rrsadykov) [reply src/Algorithm/colgenstabilization.jl:245]: Done in colgenstabilization.jl

**Pull request merged**